File: 240923/힙 정렬 구현/implement-heap-sort.js
const fs = require("fs")
const input = fs.readFileSync(0).toString().trim().split("\n")

const n = Number(input[0])
const arr = input[1].split(" ").map(x => Number(x))

function heapify(n, i)
{
    let largest = i
    let left = 2 * i
    let right = 2 * i + 1
    let temp

    if (left < n && arr[left] > arr[largest])
        largest = left
    if (right < n && arr[right] > arr[largest])
        largest = right
    
    if (largest != i)
    {
        temp = arr[largest]
        arr[largest] = arr[i]
        arr[i] = temp
        heapify(n, largest) 
    }
}

for (let i = n / 2; i >= 0; i--)
    heapify(n, i)

for (let i = n - 1; i >= 0; i--)
{
    temp = arr[0]
    arr[0] = arr[i]
    arr[i] = temp
    heapify(i, 0)
}

console.log(...arr)

File: 240923/힙 정렬 구현/implement-heap-sort.py
n = int(input())
arr = list(map(int, input().split()))

def heapify(n, i):
    largest = i
    right = i * 2
    left = i * 2 + 1

    if right < n and arr[right] > arr[largest]:
        largest = right
    if left < n and arr[left] > arr[largest]:
        largest = left

    if largest != i:
        arr[largest], arr[i] = arr[i], arr[largest]
        heapify(n, i)

for i in range(n//2, -1, -1):
    heapify(n, i)

for i in range(n-1, -1, -1):
    arr[0], arr[i] = arr[i], arr[0]
    heapify(i-1, 0)
print(*arr)